DataConnectors/Agari/AzureFunctionAgari/run.ps1

```diff
@@ -0,0 +1,380 @@
+    #Get the token for the Securty Graph API calls
+        $uri = "https://login.microsoftonline.com/$GraphTenantId/oauth2/v2.0/token"
+        $body = @{
+            client_id     = $GraphClientId
+            scope         = "https://graph.microsoft.com/.default"
+            client_secret = $GraphClientSecret
+            grant_type    = "client_credentials"
+        }
+        # Get OAuth 2.0 Token
+        $tokenRequest = Invoke-WebRequest -Method Post -Uri $uri -ContentType "application/x-www-form-urlencoded" -Body $body -UseBasicParsing
+        # Unpack Access Token if we get a success
+        if ($tokenRequest.StatusCode -eq 200) {
+            $sgapi_token = ($tokenRequest.Content | ConvertFrom-Json).access_token
+            # Base URL
+            $sgapi_uri = 'https://graph.microsoft.com/beta/security/tiIndicators'
+            $sgapi_headers = @{
+                            'Authorization' = "Bearer $sgapi_token"
+                            'ContentType' = 'application/json'
+                        } 
+        } else {
+            $sgEnabled = $false
+            Write-Host "Error fetching token for the Security Graph API call"
+        }
+        
+}
+
+#Set uris for get/revoke bearer tokens
+$get_apd_token_uri='https://api.agari.com/v1/ep/token'
+$get_bp_token_uri='https://api.agari.com/v1/cp/oauth/token'
+$get_apr_token_uri='https://api.agari.com/v1/apr/token'
+$revoke_apd_token_uri='https://api.agari.com/v1/ep/revoke'
+$revoke_apr_token_uri='https://api.agari.com/v1/apr/revoke'
+$revoke_bp_token_uri='https://api.agari.com/v1/cp/oauth/revoke'
+
+
+# Get the Bearer Tokens
+if ($bpEnabled){
+    #Get the bearer token
+    $bpToken = GetToken $client_id $client_secret $get_bp_token_uri
+    # set the headers
+    if ($bpToken){
+        $bpHeaders = @{
+                    'Authorization' = "Bearer $bpToken"
+                    'ContentType' = 'application/json'
+                }
+    } else { 
+        $bpToken = $false
+        Write-Host "Error Fetching BP Token" 
+        }
+} 
+if ($adEnabled){
+    #Get the bearer token
+    $apdToken = GetToken $client_id $client_secret $get_apd_token_uri
+    # set the headers
+    if ($apdToken){
+        $apdHeaders = @{
+                    'Authorization'="Bearer $apdToken"
+                    'ContentType' = 'application/json'
+                }
+    } else { 
+        $apdToken = $false
+        Write-Host "Error Fetching APD Token" 
+        }
+}
+if ($aprEnabled){
+    #Get the bearer token
+    $aprToken = GetToken $client_id $client_secret $get_apr_token_uri
+    # set the headers
+    if ($aprToken){
+        $aprHeaders = @{
+                    'Authorization'="Bearer $aprToken"
+                    'ContentType' = 'application/json'
+                }
+    } else { 
+        $aprToken = $false
+        Write-Host "Error Fetching APR Token" 
+        }
+}
+#Set global api variables
+$startdate=(Get-Date (get-date).addMinutes(-6) -UFormat "+%Y-%m-%dT%H:%M:01.00Z")
```

Comment:
This fetching logic is quite fragile, 

If any web request failed, the log for that 5 mins will be lost. 
Plus, according to azure function spec, the function will be triggered on every min divided by 5, this approach will for sure have duplicate data because no one can start a function at a precise time. 

In general, function trigger cannot be used to slice the log data [The trigger can be always ahead/behind schedule]. 
Check this:
https://docs.microsoft.com/en-us/azure/azure-functions/functions-bindings-timer?tabs=csharp#example

Reply:
We will accept the code as it is. We will try to provide a template to solve such issue in the future.